boot/assets.go

```diff
@@ -632,3 +632,103 @@ func (o *TrustedAssetsUpdateObserver) Canceled() error {
 	// - reseal with a given state of modeenv
 	return nil
 }
+
+func observeSuccessfulBootWithAssetsForBootloader(m *Modeenv, root string, opts *bootloader.Options) (drop []*trackedAsset, err error) {
+	trustedAssetsMap := &m.CurrentTrustedBootAssets
+	otherTrustedAssetsMap := m.CurrentTrustedRecoveryBootAssets
+	if opts != nil && opts.Role == bootloader.RoleRecovery {
+		trustedAssetsMap = &m.CurrentTrustedRecoveryBootAssets
+		otherTrustedAssetsMap = m.CurrentTrustedBootAssets
+	}
+
+	if len(*trustedAssetsMap) == 0 {
+		// bootloader may have trusted assets, but we are not tracking
+		// any for the boot process
+		return nil, nil
+	}
+
+	// let's find the bootloader first
+	bl, trustedAssets, err := findMaybeTrustedAssetsBootloader(root, opts)
+	if err != nil {
+		return nil, err
+	}
+	if len(trustedAssets) == 0 {
+		// not a trusted assets bootloader, nothing to do
+		return nil, nil
+	}
+
+	cache := newTrustedAssetsCache(dirs.SnapBootAssetsDir)
+	for _, trustedAsset := range trustedAssets {
+		assetName := filepath.Base(trustedAsset)
+
+		// find the hash of the file on disk
+		assetHash, err := cache.fileHash(filepath.Join(root, trustedAsset))
+		if err != nil && !os.IsNotExist(err) {
+			return nil, fmt.Errorf("cannot calculate the digest of existing trusted asset: %v", err)
+		}
+		if assetHash == "" {
+			// no trusted asset on disk, but we booted nonetheless
+
+			// given that asset names cannot be reused, clear the
+			// boot assets map for the current bootloader
+			delete(*trustedAssetsMap, assetName)
+			continue
+		}
+
+		// this is what we booted with
+		bootedWith := []string{assetHash}
+		// one of these was expected during boot
+		hashList := (*trustedAssetsMap)[assetName]
```

Comment:
if bootedWith is not in hashList we have a potential security issue, or some corruption, we should at least error with a potential todo to rewrite what is in the partitions with something from the cache